pkg/cloud/services/compute/instance.go

```diff
@@ -46,57 +46,150 @@ const (
-func (s *Service) constructNetworks(openStackCluster *infrav1.OpenStackCluster, instanceSpec *InstanceSpec) ([]infrav1.Network, error) {
-	trunkRequired := false
-
-	nets, err := s.getServerNetworks(instanceSpec.Networks)
-	if err != nil {
-		return nil, err
-	}
+func (s *Service) normalizePorts(ports []infrav1.PortOpts, openStackCluster *infrav1.OpenStackCluster, instanceSpec *InstanceSpec) ([]infrav1.PortOpts, error) {
+	normalizedPorts := make([]infrav1.PortOpts, 0, len(ports))
 
-	for i := range instanceSpec.Ports {
-		port := &instanceSpec.Ports[i]
-		// No Trunk field specified for the port, inherit openStackMachine.Spec.Trunk.
+	for portIdx, port := range ports {
+		// No Trunk field specified for the port, inherit the machine default
 		if port.Trunk == nil {
 			port.Trunk = &instanceSpec.Trunk
 		}
-		if *port.Trunk {
-			trunkRequired = true
-		}
-		if port.Network != nil {
-			netID := port.Network.ID
-			if netID == "" {
+
+		// Treat no Network and empty Network the same
+		noNetwork := port.Network == nil || (*port.Network == infrav1.NetworkFilter{})
+
+		// Ensure that the port has a network with an ID
+		switch {
+		// No network or subnets defined: use cluster defaults
+		case noNetwork && len(port.FixedIPs) == 0:
+			port.Network = &infrav1.NetworkFilter{
+				ID: openStackCluster.Status.Network.ID,
+			}
+			port.FixedIPs = []infrav1.FixedIP{
+				{
+					Subnet: &infrav1.SubnetFilter{
+						ID: openStackCluster.Status.Network.Subnet.ID,
+					},
+				},
+			}
+
+		// No network, but fixed IPs are defined(we handled the no fixed
+		// IPs case above): try to infer network from a subnet
+		case noNetwork:
+			s.scope.Logger().V(4).Info("No network defined for port %d, attempting to infer from subnet", portIdx)
+
+			// Look for a unique subnet defined in FixedIPs.  If we
+			// find one we can use it to infer the network ID. We
+			// don't need to worry here about the case where
+			// different FixedIPs have different networks because
+			// that will cause an error later when we try to create
+			// the port.
+			networkID, err := func() (string, error) {
 				networkingService, err := s.getNetworkingService()
 				if err != nil {
-					return nil, err
+					return "", err
 				}
 
-				netIDs, err := networkingService.GetNetworkIDsByFilter(port.Network.ToListOpt())
-				if err != nil {
-					return nil, err
+				for i, fixedIP := range port.FixedIPs {
+					if fixedIP.Subnet == nil {
+						continue
+					}
+
+					subnetOpts := fixedIP.Subnet.ToListOpt()
+					subnets, err := networkingService.GetSubnetsByFilter(&subnetOpts)
+					if err != nil {
+						return "", err
+					}
+					for len(subnets) > 1 {
+						s.scope.Logger().V(4).Info("Can't infer network from subnet %d: multiple subnets match filter", i)
+						continue
+					}
+					for _, subnet := range subnets {
+						s.scope.Logger().V(4).Info("Inferred network %s from subnet %s", subnet.NetworkID, subnet.ID)
+						return subnet.NetworkID, nil
+					}
```

Comment:
Feels like a pattern we often have, I'd at least extract it into a named function for code clarity.

Reply:
I started pulling this out, btw. I kept the non-related parts in a separate commit.